draft-ietf-oauth-transaction-tokens.md

```diff
@@ -143,6 +150,8 @@ Txn-Tokens are typically created when a workload is invoked using an endpoint th
 
 The Txn-Token Service responds to a successful invocation by generating a Txn-Token. The calling workload then uses the Txn-Token to authorize its calls to subsequent workloads. Subsequent workloads may obtain Txn-Tokens of their own.
 
+If the requesting service does not have an inbound token that it can use in its request to the Txn-Token Service, it generates a self-signed JWT and passes that in the request in place of the external authorization token.
+
```

Comment:
Are there any RFCs that gives guidance on how to generate a self-signed JWT that we can reference? For example, can we reference RFC 7523?

Reply:
I didn't see any RFCs specific to self-signed JWTs. Referring to RFC7523 is probably not appropriate because of the additional constraints that RFC puts on the content of the JWT, which may not be what we intend here.

Reply:
The self-signed JWT should be short-living as well. So "iat" and "exp" should be included. there should probably some guidance on how short those self-sign JWTs should be valid. Could be same as for Txn-Tokens itself or even shorter as the self-signed JWTs here are meant to be used just once. "aud" and "sub" should be essential with "aud" either defining the trust domain (as described elsewhere in the spec) or narrowing down the audience even further to just the TTS by using the StringorURI of the TTS. Providing the "iss" claim could help the TTS to look up the necessary JWKS of the signing party.

Reply:
> @bc-pi and @gffletch what are your thoughts on being more prescriptive on the format of the self-signed token? Can we leverage Section 3 of RFC 7523, or should we define our own? 

I wouldn't try to reference 7523 directly but maybe consider it (really just the client auth context) as partial inspiration for the claims that would make sense in this context of a self-signed JWT. Looks like @tulshi already started on a definition down in #self-signed-subject-token-type though and notably absent are `aud` and `exp`, which one would expect to see short of some compelling reason (that I'm not seeing).

Reply:
@bc-pi,  I considered adding `aud` and `exp` in the self-signed JWT, but here are the reasons why I decided not to:

- At a high-level, I'm thinking of the self-signed JWT in the request as a convenience rather than something that will have a lot of applicability outside of the communication between the requesting service and the TTS.
- We could set the `aud` value to be the TTS, but that is redundant. If we were to somehow hint about the `aud` value in the resulting TraT, that is determined by the TTS, so providing that in the request is inconsequential (and the requester has no control over it). It's possible that the same TTS has multiple trust domains, but there could be other ways for it to distinguish which trust domain it is acting on behalf of, such as a different token exchange endpoint URL.
- Similarly, the `exp` value could be used to indicate the expiration of the request itself, or it could be used as a hint to indicate the lifetime of the resulting TraT. Since we have already noted that a TTS may reject JWTs with an unreasonably old `iat` value, we do not need to specify an `exp` value. If we were to hint about the lifetime of the resulting TraT, I feel the requester would need to know a lot more about the transaction than a typical requester such as a general-purpose API gateway would know. I'd rather that the TTS determine the TraT's `exp` based on its policy and understanding of the actual transaction that is being requested.

Reply:
@tulshi I am not so sure about not defining the aud value. Given everything that goes on in he world of token theft, having this extra constraint can help minimise the impact of token replay (stealing a token from one workload and replaying against another transaction token endpoint and so forth). Are there scenarios where we think the workload would not know which transaction token service it will be presenting its self-signed credential to?

I also recall that we had some constraint that a transaction token cannot be valid for longer than the original token. This raises the question on what the guidance is for expiry if the exp claim is not present as in this case. Section 2.3 is somewhat silent on this, but I think this is a decision for the Authorization service (perhaps something to be made clearer in section 2.3).

Reply:
Claims like `exp`, `nbf`, `aud`, etc (also constructs like [explicit typing](https://www.rfc-editor.org/rfc/rfc8725.html#name-use-explicit-typing) fwiw) are about out the validity/security of the JWT in the intended context and also preventing misuse in unintended context. They have no bearing on whatever the output of the processes is. Any such bearing would need to be profiled and likely isn't what you'd want anyway. 

Note also that there are no normative processing requirements with iat.

Reply:
Thanks all, I've added the "aud" and "exp". Good points all (what was I thinking?). In order to minimize abuse of the self-signed JWT, I've specified that its lifetime should be just a few seconds. Since the Txn-token lifetime will exceed this, I've carved out that exception in "Txn-Token Lifetime" section (line 160)